make sure all torch._numpy tests run on CI

ev-br opened the pull request:
- Add `if __name__ == "__main__": run_tests()` stanzas to test files in `torch_np` folder so that these tests run on CI
- Skip / xfail things smoked out by this change 
- remove a stray python file which should not have been added to tests in the first place.
- fix einsum if opt_einsum is present
- add skips for older numpies

cc @mruberry @rgommers @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

ev-br committed:
WIP: add if __name__ == __main__ to top-level torch_np tests

ev-br committed:
CI/WIP: see how many tests fail on CI

$ pytest test/torch_np/test_basic.py               -*- 3 failures (good)
$ python test/run_tests -i torch_np/test_basic     -*- 3 failures (good)
$ python test/torch_np/test_basic.py --use-pytest  -*- 3 failures (good)
$ python test/torch_np/test_basic.py               -*- 1 failure (bad)

ev-br committed:
TST: add if __name__=="__main__" stanzas to numpy_tests/fft,linalg,lib

ev-br committed:
TST: remove a stray file in numpy_tests/lib

ev-br committed:
TST: add if __name__ == "__main__" stanzas to numpy_tests/core

ev-br committed:
TST: guard a test vs cuda.is_available

ev-br committed:
TST: xfail TestTakeAlongAxis::test_broadcast

ev-br committed:
TST: fix test_basic::TestExport

ev-br committed:
TST: fix pytest.mark.skipif

lezcano commented on test/torch_np/test_basic.py:
```diff
@@ -571,6 +572,7 @@ def test_ndarrays_to_tensors():
     assert isinstance(out[0][0], torch.Tensor)
 
 
+@pytest.mark.skipif(not torch.cuda.is_available(), reason="requires cuda")
```
There's a `skipIfNoCuda` (not real name, but you know) wrapper we use for this.

ev-br replied:
https://github.com/pytorch/pytorch/blob/main/test/dynamo/test_misc.py#L2391 uses `skipif(not TEST_CUDA)`, and `TEST_CUDA` is precisely the same: https://github.com/pytorch/pytorch/blob/main/torch/testing/_internal/common_utils.py#L1062. Will change.

ev-br committed:
MAINT: use TEST_CUDA constant from torch testing

ev-br committed:
MAINT: fix einsum if opt_einsum is installed

ev-br committed:
TST: guard NEP 50 tests for NumPy >= 1.24.x

ev-br committed:
TST: guard a test for from_dlpack new in NumPy 1.23

ev-br commented:
torch_np tests all run on CI, remaining CI failures do not seem related.

lezcano reviewed (approved):
Great! Thank you!

lezcano commented:
@pytorchbot merge -i

lezcano commented on test/torch_np/numpy_tests/lib/test_shape_base_.py:
```diff
@@ -94,6 +94,7 @@ def test_empty(self):
         actual = take_along_axis(a, ai, axis=1)
         assert_equal(actual.shape, ai.shape)
 
+    @pytest.mark.xfail(reason="https://github.com/pytorch/pytorch/pull/107875")
```
@guilhermeleobas if this PR lands before https://github.com/pytorch/pytorch/pull/108880#pullrequestreview-1618672157 you'll need to remove this xfail

clee2000 commented:
@pytorchbot revert -m "sorry but I think the asan test_scalarmath failure is real https://hud.pytorch.org/pytorch/pytorch/commit/7abeb92796635bd3ee216a0872bddd0395e97d10 https://github.com/pytorch/pytorch/actions/runs/6132913963/job/16645381921" -c ignoredsignal

ev-br committed:
TST: skip the test which triggers the integer overflow UB

Under ASAN, this crashes with

2023-09-11T23:10:59.0824026Z torch_np/numpy_tests/core/test_scalarmath.py::test_scalar_signed_integer_overflow[neg-i] /var/lib/jenkins/workspace/aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:290:3: runtime error: negation of -2147483648 cannot be represented in type 'scalar_t' (aka 'int'); cast to an unsigned type to negate this value to itself
2023-09-11T23:10:59.0825372Z     #0 0x7f86347119f4 in void at::native::AVX2::execute_op<at::native::AVX2::neg_kernel(at::TensorIteratorBase&)::$_25::operator()() const::'lambda1'()::operator()() const::'lambda'(int)&, (void*)0>(char* restrict*, long const*, long, long, at::native::AVX2::neg_kernel(at::TensorIteratorBase&)::$_25::operator()() const::'lambda1'()::operator()() const::'lambda'(int)&) (/opt/conda/envs/py_3.9/lib/python3.9/site-packages/torch/lib/libtorch_cpu.so+0x140f49f4)
2023-09-11T23:10:59.0826756Z     #1 0x7f863471164d in void at::native::AVX2::basic_loop<at::native::AVX2::neg_kernel(at::TensorIteratorBase&)::$_25::operator()() const::'lambda1'()::operator()() const::'lambda'(int)&>(char* restrict*, long const*, long, long, at::native::AVX2::neg_kernel(at::TensorIteratorBase&)::$_25::operator()() const::'lambda1'()::operator()() const::'lambda'(int)&) (/opt/conda/envs/py_3.9/lib/python3.9/site-packages/torch/lib/libtorch_cpu.so+0x140f464d)
2023-09-11T23:10:59.0828284Z     #2 0x7f8634711e3e in at::native::AVX2::VectorizedLoop2d<at::native::AVX2::neg_kernel(at::TensorIteratorBase&)::$_25::operator()() const::'lambda1'()::operator()() const::'lambda

The test was actually xfailed, skip it instead to avoid even running it.

ev-br commented:
The offending test was doing essentially this:

```
(Pdb) p x
tensor(-2147483648, dtype=torch.int32)
(Pdb) p torch.iinfo(torch.int32).min
-2147483648
(Pdb) p torch.negative(x)
tensor(-2147483648, dtype=torch.int32)      # NB: equals `x`
(Pdb) p torch.negative(x.to(torch.int64))
tensor(2147483648)
```

under ASAN, this produced instead (from the CI logs):

```
    2023-09-11T23:10:59.0824026Z torch_np/numpy_tests/core/test_scalarmath.py::test_scalar_signed_integer_overflow[neg-i] /var/lib/jenkins/workspace/aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:290:3: runtime error: negation of -2147483648 cannot be represented in type 'scalar_t' (aka 'int'); cast to an unsigned type to negate this value to itself
    2023-09-11T23:10:59.0825372Z     #0 0x7f86347119f4 in void at::native::AVX2::execute_op<at::native::AVX2::neg_kernel(at::TensorIteratorBase&)::$_25::operator()() const::'lambda1'()::operator()() const::'lambda'(int)&, (void*)0>(char* restrict*, long const*, long, long, at::native::AVX2::neg_kernel(at::TensorIteratorBase&)::$_25::operator()() const::'lambda1'()::operator()() const::'lambda'(int)&) (/opt/conda/envs/py_3.9/lib/python3.9/site-packages/torch/lib/libtorch_cpu.so+0x140f49f4)
    2023-09-11T23:10:59.0826756Z     #1 0x7f863471164d in void at::native::AVX2::basic_loop<at::native::AVX2::neg_kernel(at::TensorIteratorBase&)::$_25::operator()() const::'lambda1'()::operator()() const::'lambda'(int)&>(char* restrict*, long const*, long, long, at::native::AVX2::neg_kernel(at::TensorIteratorBase&)::$_25::operator()() const::'lambda1'()::operator()() const::'lambda'(int)&) (/opt/conda/envs/py_3.9/lib/python3.9/site-packages/torch/lib/libtorch_cpu.so+0x140f464d)
    2023-09-11T23:10:59.0828284Z     #2 0x7f8634711e3e in at::native::AVX2::VectorizedLoop2d<at::native::AVX2::neg_kernel(at::TensorIteratorBase&)::$_25::operator()() const::'lambda1'()::operator()() const::'lambda
```

While the operation is actually UB in C, the python-level test:
- in numpy it checked that a RuntimeWarning is emitted on overflow
- in torch._numpy we xfailed it, since pytorch does not emit these warnings.

An xfailed test still runs, hence the CI failure. Therefore, the last commit changes `pytest.mark.xfail` to `pytest.mark.skip` --- let's see if it's enough to prevent it from running at all.

ev-br commented:
@pytorchbot merge -i
